### DISCUSSION TITLE
Why is `maybe2` needed?

### DISCUSSION THREAD

**Discussion by osa1** [koka-lang/koka#627] [Q&A] [answered]
Since `tuple2` and `maybe` are both value types, a type like `maybe2<int, int>` should be the same as `maybe<tuple2<int, int>>`, no? Why do we need `maybe2`?

https://github.com/koka-lang/koka/blob/21089c158ef19dbecf0a3d934a04b8ecbff20125/lib/std/core/types.kk#L204-L208

**Comment by TimWhiting** [comment, marked as answer]
I could be wrong, but my understanding is that value types are themselves stored as C structs, but any polymorphic fields are boxed.
Thus a tuple is represented by a C struct containing two boxed fields, and the maybe type is a C struct containing the constructor tag and a union of nothing and the field for the just constructor (i.e. a tagged union). However, a tuple in order to be stored in the just constructor would involve the tuple itself needing to be boxed to fit in the polymorphic field (which is just a pointer). Koka doesn't specialize datatypes currently and I know of no plans for that.

TLDR; value types are not entirely value types in the presence of polymorphism. They don't need heap allocation for the datatype themself if used directly, but do need boxing to fit into polymorphic containers. (And the same goes for higher order function parameters that are polymorphic).

**Reply by osa1** [reply]
Ah, of course. We need monomorphisation if we want `maybe<x>` to store `x` boxed when it's not a value type, and unboxed when it is, and Koka doesn't do monomorphisation.
